src/v2/features/characters/characters-value.service.ts

```diff
@@ -14,109 +13,103 @@ export class CharactersValueService {
         }
 
         const valueLevel =
-            // CharactersValueService.getUnlockValue(character.initialRarity, character.name) +
-            // CharactersValueService.getExperienceValue(character.level) +
-            // CharactersValueService.getAbilityValue(character.activeAbilityLevel) +
-            // CharactersValueService.getAbilityValue(character.passiveAbilityLevel) +
-            // CharactersValueService.getStarsValue(character.stars) -
-            // CharactersValueService.getInitialStarsValue(character.initialRarity) +
-            // CharactersValueService.getRarityValue(character.rarity) -
-            // CharactersValueService.getRarityValue(character.initialRarity) +
-            // not yet working:
+            CharactersValueService.getUnlockValue(character.initialRarity, character.name) +
+            CharactersValueService.getExperienceValue(character.level) +
+            CharactersValueService.getAbilityValue(character.activeAbilityLevel) +
+            CharactersValueService.getAbilityValue(character.passiveAbilityLevel) +
+            CharactersValueService.getStarsValue(character.stars) -
+            CharactersValueService.getInitialStarsValue(character.initialRarity) +
+            CharactersValueService.getRarityValue(character.rarity) -
+            CharactersValueService.getRarityValue(character.initialRarity) +
             CharactersValueService.getRankValue(character.name, character.rank, character.upgrades);
         return Math.round(valueLevel);
     }
 
+    public static getUpgrades(...characters: Array<ICharacterRankRange>): IMaterialFull[] {
+        const rankEntries: number[] = getEnumValues(Rank).filter(x => x > 0);
+        const result: IMaterialFull[] = [];
+        let priority = 0;
+        for (const character of characters) {
+            priority++;
+            const characterUpgrades = StaticDataService.rankUpData[character.id];
+            if (!characterUpgrades) {
+                continue;
+            }
+            const ranksRange = rankEntries.filter(r => r >= character.rankStart && r < character.rankEnd);
+
+            const rankUpgrades = ranksRange
+                .flatMap((rank, index) => {
+                    const result = characterUpgrades[rankToString(rank)] ?? [];
+                    return index === 0 ? result.filter(x => !character.appliedUpgrades.includes(x)) : result;
```

Comment:
Include upgrades from the last (current rank).
`return index ===  ranksRange.length - 1 ? result.filter(x => character.appliedUpgrades.includes(x)) : result;`